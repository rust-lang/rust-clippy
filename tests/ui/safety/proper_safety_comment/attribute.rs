#![warn(clippy::proper_safety_comment)]

//~v ERROR: missing safety comment on critical attribute
#[unsafe(no_mangle)]
pub fn f1() {}

//~v ERROR: unnecessary safety comment on attribute
// SAFETY: ...
#[warn(clippy::proper_safety_comment)]
#[unsafe(no_mangle)]
//~^ ERROR: missing safety comment on critical attribute
pub fn f2() {}

#[warn(clippy::proper_safety_comment)]
// SAFETY: ...
#[unsafe(no_mangle)]
pub fn f3() {}

fn nested() {
    //~v ERROR: missing safety comment on critical attribute
    #[unsafe(no_mangle)]
    pub fn f4() {}
}

// SAFETY: not detected as unnecessary safety comment due to the procedural macro
#[derive(Debug)]
struct S1;

#[derive(Debug)]
struct S2;

fn main() {}
