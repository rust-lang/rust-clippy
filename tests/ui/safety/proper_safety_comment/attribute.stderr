error: missing safety comment on critical attribute
  --> tests/ui/safety/proper_safety_comment/attribute.rs:4:1
   |
LL | #[unsafe(no_mangle)]
   | ^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::proper-safety-comment` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::proper_safety_comment)]`

error: unnecessary safety comment on attribute
  --> tests/ui/safety/proper_safety_comment/attribute.rs:8:1
   |
LL | / // SAFETY: ...
LL | | #[warn(clippy::proper_safety_comment)]
   | |_^

error: missing safety comment on critical attribute
  --> tests/ui/safety/proper_safety_comment/attribute.rs:10:1
   |
LL | #[unsafe(no_mangle)]
   | ^^^^^^^^^^^^^^^^^^^^

error: missing safety comment on critical attribute
  --> tests/ui/safety/proper_safety_comment/attribute.rs:21:5
   |
LL |     #[unsafe(no_mangle)]
   |     ^^^^^^^^^^^^^^^^^^^^

error: aborting due to 4 previous errors

